erdos/src/dataflow/state.rs

```diff
@@ -1,43 +1,92 @@
-
-/// Trait that must be implemented by a state structure that is used in a Parallel operator.
-/// This structure must implement a `commit` method that commits the final state for a given
-/// timestamp `t`.
-pub trait StateT: 'static + Send + Sync {
+    /// The `commit` method commits the final state for a given timestamp.
     fn commit(&mut self, timestamp: &Timestamp);
 
-    fn get_last_committed_timestamp(&self) -> Timestamp;
+    /// Retrieves the last committed timestamp by this state.
+    /// This method can be used in conjunction with `at` to retrieve the latest committed state.
+    fn last_committed_timestamp(&self) -> Timestamp;
+
+    /// Retrieve the state at a given timestamp.
+    /// If the state for that timestamp hasn't been initialized yet, invoke the default method on
+    /// the `Item` type, and return the newly created state for that timestamp.
+    fn at(&mut self, timestamp: &Timestamp) -> Option<&mut Self::Item>;
 }
 
-impl StateT for () {
+/// State implementation for () to be used by operators that are stateless.
+impl State for () {
+    type Item = ();
+
     fn commit(&mut self, _timestamp: &Timestamp) {}
 
-    fn get_last_committed_timestamp(&self) -> Timestamp {
+    fn last_committed_timestamp(&self) -> Timestamp {
         Timestamp::Bottom
     }
+
+    fn at(&mut self, _timestamp: &Timestamp) -> Option<&mut Self::Item> {
+        None
+    }
+}
+
+/// The `TimeVersionedState` provides a default implementation of the `State` for a type S.
+/// The structure automatically commits the final state for a timestamp into a HashMap and
+/// initializes new states for a timestamp `t` by invoking their default method.
+pub struct TimeVersionedState<S>
+where
+    S: 'static + Default + Send + Sync,
+{
+    state: HashMap<Timestamp, S>,
+    last_committed_timestamp: Timestamp,
+}
+
+impl<S> TimeVersionedState<S>
+where
+    S: 'static + Default + Send + Sync,
+{
+    pub fn new() -> Self {
+        Self {
+            state: HashMap::new(),
+            last_committed_timestamp: Timestamp::Bottom,
+        }
+    }
+}
+
+impl<S> State for TimeVersionedState<S>
+where
+    S: 'static + Default + Send + Sync,
+{
+    type Item = S;
+
+    fn commit(&mut self, timestamp: &Timestamp) {
+        self.last_committed_timestamp = timestamp.clone();
+    }
+
+    fn last_committed_timestamp(&self) -> Timestamp {
+        self.last_committed_timestamp.clone()
+    }
+
+    fn at(&mut self, timestamp: &Timestamp) -> Option<&mut Self::Item> {
+        Some(self.state.entry(timestamp.clone()).or_default())
+    }
 }
 
 /// Trait that must be implemented by a state structure that is used in a Sequential operator.
 /// This state structure must implement an `append` method that enables message callbacks to add
 /// intermediate state to the structure, and a `commit` method that commits the final state for a
 /// given timestamp t.
```

Comment:
Could you move the documentation of the methods above the method declarations?

E.g.
```rust
   /// Adds intermediate state to the structure. Invoked by a callback.
   fn append(&self, data &S);
```